src/server/pfs/server/driver_new_storage.go

```diff
@@ -298,24 +299,105 @@ func (fr *FileReader) drain() error {
 	return nil
 }
 
-func (d *driver) compact(master *work.Master, outputPath string, inputPrefixes []string) (retErr error) {
+// compact is the entrypoint to compaction.
+func (d *driver) compact(ctx context.Context, master *work.Master, outputPath string, inputPrefixes []string) error {
+	// resolve prefixes into paths
+	inputPaths := []string{}
+	for _, inputPrefix := range inputPrefixes {
+		err := d.storage.WalkFileSets(ctx, inputPrefix, func(inputPath string) error {
+			inputPaths = append(inputPaths, inputPath)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+	if d.env.StorageCompactionMaxFanIn < 2 {
+		panic("StorageCompactionMaxFanIn cannot be < 2")
+	}
+	return d.compactIter(ctx, compactParams{
+		master:     master,
+		inputPaths: inputPaths,
+		outputPath: outputPath,
+		maxFanIn:   d.env.StorageCompactionMaxFanIn,
+	})
+}
+
+type compactParams struct {
```

Comment:
I think `compactSpec` might be a better name.

Reply:
Isn't there a method `CompactSpec` on the storage type that would get confused with this?

Reply:
Yeah, that is actually why I prefer it. There is a notion of a fileset compact spec and a pfs compact spec. The fileset compact spec shards across a set of filesets, and the pfs compact spec manages fan out.